sdk/core/Azure.Core.Experimental/tests/JsonDataChangeListTests.cs

```diff
@@ -0,0 +1,301 @@
+            Assert.AreEqual(5.0, baz.Baz.B);
+        }
+
+        private class BazA
+        {
+            public double Foo { get; set; }
+            public A_ Baz { get; set; }
+        }
+
+        private class BazB
+        {
+            public double Foo { get; set; }
+            public B_ Baz { get; set; }
+        }
+
+        private class A_
+        {
+            public double A { get; set; }
+        }
+
+        private class B_
+        {
+            public double B { get; set; }
+        }
+
+        [Test]
+        public void CanGetArrayElement()
+        {
+            string json = @"
+                {
+                  ""Foo"" : [ 1, 2, 3 ]
+                }";
+
+            var jd = JsonData.Parse(json);
+
+            Assert.AreEqual(1, jd.RootElement.GetProperty("Foo").GetIndexElement(0).GetInt32());
+            Assert.AreEqual(2, jd.RootElement.GetProperty("Foo").GetIndexElement(1).GetInt32());
+            Assert.AreEqual(3, jd.RootElement.GetProperty("Foo").GetIndexElement(2).GetInt32());
+        }
+
+        [Test]
+        public void CanSetArrayElement()
+        {
+            string json = @"
+                {
+                  ""Foo"" : [ 1, 2, 3 ]
+                }";
+
+            var jd = JsonData.Parse(json);
+
+            jd.RootElement.GetProperty("Foo").GetIndexElement(0).Set(5);
+            jd.RootElement.GetProperty("Foo").GetIndexElement(1).Set(6);
+            jd.RootElement.GetProperty("Foo").GetIndexElement(2).Set(7);
+
+            Assert.AreEqual(5, jd.RootElement.GetProperty("Foo").GetIndexElement(0).GetInt32());
+            Assert.AreEqual(6, jd.RootElement.GetProperty("Foo").GetIndexElement(1).GetInt32());
+            Assert.AreEqual(7, jd.RootElement.GetProperty("Foo").GetIndexElement(2).GetInt32());
+        }
+
+        [Test]
+        public void CanSetArrayElementMultipleTimes()
+        {
+            string json = @"
+                {
+                  ""Foo"" : [ 1, 2, 3 ]
+                }";
+
+            var jd = JsonData.Parse(json);
+
+            jd.RootElement.GetProperty("Foo").GetIndexElement(0).Set(5);
+            jd.RootElement.GetProperty("Foo").GetIndexElement(0).Set(6);
+
+            Assert.AreEqual(6, jd.RootElement.GetProperty("Foo").GetIndexElement(0).GetInt32());
+        }
+
+        [Test]
+        public void CanSwapArrayElements()
+        {
+            string json = @"[ { ""Foo"" : 4 } ]";
+
+            var jd = JsonData.Parse(json);
+
+            var a = jd.RootElement.GetIndexElement(0);
+            jd.RootElement.GetIndexElement(0).Set(5);
+
+            // This is wicked because 'a' keeps a reference to the root
+            // with the changelist.  Would we detach that?  How would we know to?
+            a.GetProperty("Foo").Set(6);
+
+            Assert.AreEqual(5, jd.RootElement.GetIndexElement(0).GetInt32());
+            Assert.AreEqual(6, a.GetProperty("Foo").GetIndexElement(0).GetInt32());
```

Comment:
line 260 set `a.Foo` to the number 6. Numbers don't have indexers so a.Foo[0] throws, as it should.

Reply:
I changed the line above to `Assert.AreEqual(6, a.GetProperty("Foo").GetInt32());` and my test passes

Reply:
The original pseudocode we were discussing was the following:

```csharp
var json = DynamicJson.Parse("""[ { "Bar" : 4 } ]""");
dynamic a = json[0]; 
json[0] = 5;
a.Bar = 6;
Console.WriteLine(json[0]);
Console.WriteLine(a);
json[0] = a;
Console.WriteLine(json[0]);
Console.WriteLine(json.ToPatch());
```

I had interpreted this as `a` is now a JsonDataElement representing the JSON object `{"Bar":4}`, even if `json[0]` is reassigned in the original object.  But I'm happy to discuss the interpretation if you think these aren't the right semantics for the type.

Reply:
Addressed, and tested in the reference semantics tests in `MutableJsonDocumentTests`